### PR TITLE
[google_maps_flutter] Add examples for different iOS versions

### PR DESCRIPTION
**Pull request by stuartmorgan**
Improves the way we specify and test the `GoogleMaps` SDK dependency on iOS. Currently as discussed in https://github.com/flutter/flutter/issues/86820, we are allowing any version of the SDK, and CocoaPods will choose the best version that's compatible with the minimum iOS deployment version of the plugin client's app. However, clients are also subject to future breakage, since any new major SDK version could have compile-time incompatibility with the plugin's wrapping code. We also have no ability to test anything in CI that's newer than the SDK that supports our oldest supported version (currently that's iOS 11, and thus 5.x).

This makes two changes:
- Replaces the single example with N examples, each with a different minimum iOS version, corresponding to each point where the Google Maps SDK has bumped its minimum iOS requirement. Currently (as documented in the new README file)
  - The oldest example has all of the current testing.
  - The newest example has just XCTests (not integration or XCUITests).
  - Intervening examples have no tests.
- Instead of the completely unpinned version, we allow anything up to the next major version (which should be the first one with breaking API changes).

This gives us build coverage of each of the (latest at CI runtime) resolve points that clients can have. We retain the flexibility of allowing clients to get the best version that meets their app's constraints, except that we will need to explicitly test and enable new major versions, so that they won't randomly break clients.

Fixes https://github.com/flutter/flutter/issues/86820

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
This looks like a huge review, but the ios12/ and ios13/ folders are just exact copies of the ios11 folder with:
- the minimum deployment version changed, and
- most of the tests removed.

**Review by cyanglaz — APPROVED** [review]
LGTM

**Review by jmagman — COMMENTED** [review]
1. How is this building/running tests?  Or is it?  The flutter tool doesn't know anything about an `ios11` directory but would be looking for an `ios` directory.
2. Is there a way we can do this that just updates the minimum version instead of duplicating everything needed for the platform directory? For example, keep the single `ios` directory but have 3 app targets with different versions?  Or set an environment variable the Podfile can parse like (I didn't test this):
```ruby
min_ios_version = ENV["MINIMUM_IOS_VERSION"] || '11.0'
platform :ios, min_ios_version
```
plus maybe an xcconfig that can be set for the `IPHONEOS_DEPLOYMENT_TARGET`?

**Comment by stuartmorgan** [comment]
> 1. How is this building/running tests?  Or is it?  The flutter tool doesn't know anything about an `ios11` directory but would be looking for an `ios` directory.

`ios11` isn't a platform directory, it's a whole project directory. This works because the repo tooling is already set up to understand two different structures for examples:
- example/, a Flutter app project
- example/, a directory containing any number of directories with arbitrary names which are themselves distinct example projects

The tooling sees that there are three examples, and reacts accordingly. For `build-examples` it builds all three. For `drive-examples` it runs the one with integration tests and skips the others (and passes because the tooling just requires that there be at least one example with integration tests, not that all of them have one). For `native-test` it runs the ones with native tests (currently ios11 and ios13), and skips the one without any test targets.

> 2. Is there a way we can do this that just updates the minimum version instead of duplicating everything needed for the platform directory? For example, keep the single `ios` directory but have 3 app targets with different versions?  Or set an environment variable the Podfile can parse like (I didn't test this):
> 
> ```ruby
> min_ios_version = ENV["MINIMUM_IOS_VERSION"] || '11.0'
> platform :ios, min_ios_version
> ```
> 
> plus maybe an xcconfig that can be set for the `IPHONEOS_DEPLOYMENT_TARGET`?

We'd have to build bespoke tooling that knew how to set all of this up, and also run different sets of tests for each scenario (e.g., we don't want to run integration tests three times because it would add a bunch of CI time for minimal benefit). It's doable, but it seems like a lot more trouble than just having several copies of files that are almost all just boilerplate.

I could extract the Dart for the example to a shared path-based package, so we don't have three copies of that. That would be easy, and since that part isn't boilerplate it would reduce maintenance if we need to update the examples. I was going back and forth on it because it'll make the structure a little weirder, but it's probably better on the whole to have the one copy.

**Review by jmagman — APPROVED** [review]
Thanks for the explanation @stuartmorgan, and for getting these tested.  LGTM

File: packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+of this platform implementation. It is not intended to demonstrate actual use of
+this package, since the intent is that plugin clients use the app-facing
+package.
+
+Unless you are making changes to this implementation package, this example is
+very unlikely to be relevant.
+
+## Versions
+
+This example requires iOS 11, so will select a 5.x SDK version.
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/README.md]: Oops, I forget about this. Fixed in all three readmes.

**Pull request merged**